### ISSUE TITLE
关于使用其他模型进行训练的问题

### ISSUE THREAD

**Issue by Tunanzzz** [Kwai-YuanQi/MM-RLHF#7]
你好，

首先，感谢你们出色的工作，给了我很大的帮助和启发~

我注意到文件目录里面有swift相关的sh脚本，请问这个项目是基于swift构建的吗？如果我希望对其他模型进行微调例如Qwen2.5VL-7B或者Intern3VL-8B等，我应该怎么做呢？

期待你们的回复

**Comment by Tunanzzz** [comment]
如果能够是在swift文件目录下直接修改dpo的训练方式将会是更好的帮助

**Comment by yfzhang114** [comment]
这个项目不是基于swfit构建的，但是我们的数据可以很容易适配到swift的格式，通过scripts/convert_to_swift.py，你可以使用swift的任意算法在我们的数据上进行训练。

**Comment by Tunanzzz** [comment]
感谢你的回复~

因为文章对dpo算法有一定改进，我如果用swift项目来训练其他的模型的话，应该修改哪些文件呢？

**Comment by yfzhang114** [comment]
loss直接修改swift/trainers/rlhf_trainer/dpo_trainer.py即可，输入模型的input需要找到对应位置进行更改

**Comment by Tunanzzz** [comment]
能具体一些吗T.T，或者如果能在MM-RLHF项目中适配Qwen2.5VL模型也可以

**Comment by Tunanzzz** [comment]
在Readme.md中提到的：“MM-RLHF is now fully supported by [swift](https://github.com/modelscope/ms-swift). Simply process your data using scripts/convert_to_swift.py and execute the scripts/swift.sh command to get started.”，在不对ms-swift的代码进行任何修改的前提下，这样直接启用swift进行dpo是不是没有应用本文提出的mm-dpo的损失函数方式，利用reward信息

**Comment by yfzhang114** [comment]
你的理解是对的

**Comment by Tunanzzz** [comment]
我希望用Qwen2.5VL来训练的话，具体应该怎么做呢

**Comment by Tunanzzz** [comment]
swift中的dpo_trainer.py的内容为：
`# Copyright (c) Alibaba, Inc. and its affiliates.
from typing import Dict, List, Optional, Tuple, Union

import torch
import torch.nn as nn
from peft import PeftModel
from transformers import PreTrainedModel
from trl import DPOTrainer as HFDPOTrainer

from ..mixin import DataLoaderMixin, SwiftMixin
from .rlhf_mixin import RLHFTrainerMixin

del HFDPOTrainer.__init__


class DPOTrainer(RLHFTrainerMixin, SwiftMixin, DataLoaderMixin, HFDPOTrainer):

    def __init__(self,
                 model: Optional[Union[PreTrainedModel, nn.Module, str]] = None,
                 ref_model: Optional[Union[PreTrainedModel, nn.Module, str]] = None,
                 *_args,
                 **kwargs):
        from trl.trainer import FDivergenceConstants
        args = kwargs['args']
        self.label_smoothing = args.label_smoothing
        self.loss_type = args.loss_type
        self.precompute_ref_log_probs = args.precompute_ref_log_probs
        self.f_divergence_type = args.f_divergence_type
        self.f_divergence_params = {FDivergenceConstants.ALPHA_DIVERGENCE_COEF_KEY: args.f_alpha_divergence_coef}
        self.is_peft_model = isinstance(model, PeftModel)

        self.ref_adapter_name = args.ref_adapter_name
        self.reference_free = args.reference_free
        self.use_weighting = False

        super().__init__(model, ref_model, *_args, **kwargs)

    def get_nll_loss(self, logits, labels):
        if not self.is_encoder_decoder:
            # Shift so that tokens < n predict n
            logits = logits[..., :-1, :].contiguous()
            labels = labels[..., 1:].contiguous()
        # Flatten the tokens
        loss_fct = nn.CrossEntropyLoss(ignore_index=self.label_pad_token_id)
        logits = logits.view(-1, logits.shape[-1])
        labels = labels.view(-1)
        # Enable model parallelism
        labels = labels.to(logits.device)
        return loss_fct(logits, labels)

    def concatenated_forward(
        self, model: nn.Module, batch: Dict[str, Union[List, torch.LongTensor]]
    ) -> Tuple[torch.FloatTensor, torch.FloatTensor, torch.FloatTensor, torch.FloatTensor, torch.FloatTensor]:
        batch = batch.copy()
        num_examples = batch['labels'].shape[0] // 2
        labels = batch.pop('labels', None)
        if self.is_encoder_decoder:
            batch['labels'] = labels

        if self.aux_loss_enabled:
            batch['output_router_logits'] = True
        outputs = model(**batch, use_cache=False)
        batch['labels'] = labels
        if outputs.logits.shape[1] != labels.shape[1]:
            # for llava, the model returns logits for the entire sequence, including the image tokens
            # (placed before the text tokens)
            outputs.logits = outputs.logits[:, -labels.shape[1]:]
        for key in ['input_ids', 'attention_mask', 'labels']:
            batch[f'concatenated_{key}'] = batch.pop(key, None)
        if self.__class__.__name__ == 'ORPOTrainer':  # Pass-through labels
            batch['concatenated_input_ids'] = batch['concatenated_labels']

        all_logits = outputs.logits

        if all_logits.shape[:2] != batch['concatenated_labels'].shape[:2]:
            # for llava, the model returns logits for the entire sequence,
            # including the image tokens (placed before the text tokens)
            seq_len = batch['concatenated_labels'].shape[1]
            all_logits = all_logits[:, -seq_len:]

        all_logps, size_completion = self.get_batch_logps(
            all_logits,
            batch['concatenated_labels'],
            is_encoder_decoder=self.is_encoder_decoder,
            label_pad_token_id=self.label_pad_token_id,
        )

        output = {}

        if self.args.rpo_alpha is not None:
            labels = batch['concatenated_labels'].clone()
            output['nll_loss'] = self.get_nll_loss(all_logits[:num_examples], labels[:num_examples])

        if self.loss_type == 'ipo':
            all_logps = all_logps / size_completion

        output['chosen_logps'] = all_logps[:num_examples]
        output['rejected_logps'] = all_logps[num_examples:]
        output['mean_chosen_logits'] = all_logits[:num_examples].mean()
        output['mean_rejected_logits'] = all_logits[num_examples:].mean()

        if self.aux_loss_enabled:
            output['aux_loss'] = outputs.aux_loss

        return output

    @staticmethod
    def get_batch_logps(
        logits: torch.FloatTensor,
        labels: torch.LongTensor,
        label_pad_token_id: int = -100,
        is_encoder_decoder: bool = False,
    ) -> Tuple[torch.FloatTensor, torch.LongTensor]:
        if logits.shape[:-1] != labels.shape:
            raise ValueError(f'Logits (batch and sequence length dim) {logits.shape[:-1]}'
                             'and labels must have the same shape {labels.shape}')
        if not is_encoder_decoder:
            labels = labels[:, 1:].clone()
            logits = logits[:, :-1, :]
        else:
            labels = labels.clone()

        loss_mask = labels != label_pad_token_id

        labels[labels == label_pad_token_id] = 0

        per_token_logps = torch.gather(logits.log_softmax(-1), dim=2, index=labels.unsqueeze(2)).squeeze(2)

        return (per_token_logps * loss_mask).sum(-1), loss_mask.sum(-1)
`

**Comment by yfzhang114** [comment]
dpo算法和backbone是没关系的，loss的修改逻辑和我们的trainer一致即可
